File: lib/src/features/plans/data/plans_repository.dart
import 'package:drift/drift.dart';
import 'package:uuid/uuid.dart';

import '../../../core/database/database.dart';
import '../../../core/database/sync.dart';
import '../../../core/database/tables.dart';
import 'plans_dao.dart';

/// Offline-first seam for plans and their timeline items. Local-only today;
/// remote read/write and sync queueing land here later.
class PlansRepository {
  PlansRepository(this._dao, {Uuid? uuid}) : _uuid = uuid ?? const Uuid();

  final PlansDao _dao;
  final Uuid _uuid;

  Stream<List<Plan>> watchPlans() => _dao.watchPlans();

  Future<Plan?> findPlan(String id) => _dao.findPlan(id);

  Stream<List<PlanItem>> watchItems(String planId) => _dao.watchItems(planId);

  /// Create (id null) or update a plan. Returns the plan's id so callers can
  /// attach items to a freshly created plan.
  Future<String> savePlan({
    String? id,
    required String name,
    required DateTime date,
    required ActivityType activityType,
    required double distanceKm,
    required int durationMinutes,
    double targetCarbsPerHour = 0,
    double targetSodiumPerHour = 0,
    double targetCaffeinePerHour = 0,
    PlanType? planType,
    double? intakeInterval,
    String? comments,
  }) async {
    final planId = id ?? _uuid.v4();
    await _dao.upsertPlan(
      PlansCompanion(
        id: Value(planId),
        name: Value(name),
        date: Value(date),
        activityType: Value(activityType),
        distanceKm: Value(distanceKm),
        durationMinutes: Value(durationMinutes),
        targetCarbsPerHour: Value(targetCarbsPerHour),
        targetSodiumPerHour: Value(targetSodiumPerHour),
        targetCaffeinePerHour: Value(targetCaffeinePerHour),
        planType: Value(planType),
        intakeInterval: Value(intakeInterval),
        comments: Value(comments),
        updatedAt: Value(DateTime.now()),
        syncStatus: const Value(SyncStatus.pending),
      ),
    );
    return planId;
  }

  /// Add (id null) or update one food entry on a plan's timeline.
  Future<void> saveItem({
    String? id,
    required String planId,
    required String foodId,
    required double offsetLength,
    double quantity = 1,
  }) {
    return _dao.upsertItem(
      PlanItemsCompanion(
        id: Value(id ?? _uuid.v4()),
        planId: Value(planId),
        foodId: Value(foodId),
        offsetLength: Value(offsetLength),
        quantity: Value(quantity),
        updatedAt: Value(DateTime.now()),
        syncStatus: const Value(SyncStatus.pending),
      ),
    );
  }

  Future<void> deletePlan(String id) => _dao.softDeletePlan(id);

  Future<void> deleteItem(String id) => _dao.softDeleteItem(id);
}

File: react-native/src/features/plans/data/plans-repository.ts
import { eq } from 'drizzle-orm';
import * as Crypto from 'expo-crypto';

import { db } from '@/core/db/client';
import {
  planItems,
  plans,
  type ActivityType,
  type Plan,
  type PlanType,
} from '@/core/db/schema';

export type PlanInput = {
  id?: string;
  name: string;
  date: Date;
  activityType: ActivityType;
  distanceKm: number;
  durationMinutes: number;
  targetCarbsPerHour?: number;
  targetSodiumPerHour?: number;
  targetCaffeinePerHour?: number;
  planType?: PlanType | null;
  intakeInterval?: number | null;
  comments?: string | null;
};

export async function findPlan(id: string): Promise<Plan | undefined> {
  const rows = await db.select().from(plans).where(eq(plans.id, id)).limit(1);
  return rows.at(0);
}

/// Create (id undefined) or update a plan. Returns the id.
export async function savePlan(input: PlanInput): Promise<string> {
  const id = input.id ?? Crypto.randomUUID();
  const now = new Date();
  const fields = {
    name: input.name,
    date: input.date,
    activityType: input.activityType,
    distanceKm: input.distanceKm,
    durationMinutes: input.durationMinutes,
    targetCarbsPerHour: input.targetCarbsPerHour ?? 0,
    targetSodiumPerHour: input.targetSodiumPerHour ?? 0,
    targetCaffeinePerHour: input.targetCaffeinePerHour ?? 0,
    planType: input.planType ?? null,
    intakeInterval: input.intakeInterval ?? null,
    comments: input.comments ?? null,
    updatedAt: now,
    syncStatus: 'pending' as const,
  };
  await db
    .insert(plans)
    .values({ id, ...fields })
    .onConflictDoUpdate({ target: plans.id, set: fields });
  return id;
}

/// Soft delete a plan and its timeline items together.
export async function deletePlan(id: string): Promise<void> {
  const now = new Date();
  const gone = { deletedAt: now, updatedAt: now, syncStatus: 'pending' as const };
  await db.update(plans).set(gone).where(eq(plans.id, id));
  await db.update(planItems).set(gone).where(eq(planItems.planId, id));
}

/// Choose how a plan's intake is tracked (by distance or time) and the
/// interval between suggested fuel points. Unlocks the timeline.
export async function setIntakeTracking(args: {
  id: string;
  planType: PlanType;
  interval: number;
}): Promise<void> {
  await db
    .update(plans)
    .set({
      planType: args.planType,
      intakeInterval: args.interval,
      updatedAt: new Date(),
      syncStatus: 'pending',
    })
    .where(eq(plans.id, args.id));
}

/// Place a food on a plan's timeline at an offset (km or min). Returns the id.
export async function saveItem(args: {
  planId: string;
  foodId: string;
  offsetLength: number;
  quantity?: number;
}): Promise<string> {
  const id = Crypto.randomUUID();
  await db.insert(planItems).values({
    id,
    planId: args.planId,
    foodId: args.foodId,
    offsetLength: args.offsetLength,
    quantity: args.quantity ?? 1,
    updatedAt: new Date(),
    syncStatus: 'pending',
  });
  return id;
}

/// Fine-tune a placed item's position and servings.
export async function updateItem(args: {
  id: string;
  offsetLength: number;
  quantity: number;
}): Promise<void> {
  await db
    .update(planItems)
    .set({
      offsetLength: args.offsetLength,
      quantity: args.quantity,
      updatedAt: new Date(),
      syncStatus: 'pending',
    })
    .where(eq(planItems.id, args.id));
}

/// Soft delete a single timeline item.
export async function deleteItem(id: string): Promise<void> {
  const now = new Date();
  await db
    .update(planItems)
    .set({ deletedAt: now, updatedAt: now, syncStatus: 'pending' })
    .where(eq(planItems.id, id));
}
